hadoop-hdds/docs/content/design/container-reconciliation.md

```diff
@@ -0,0 +1,344 @@
+
+This read could be an end client or another datanode asking for data to repair its replica. The result will only be inconsistent if the chunk/block is being repaired. In this case the data is already corrupt and the client will get a read checksum failure either way. Even if all datanodes are reconciling the same container at the same time, this should not cause an availability problem because there has to be one correct replica that others are repairing from that the client can also read.
+
+#### Replicate command for a container being reconciled
+
+- Allow replication to proceed with best effort, possibly with a warning log.
+
+There may be cases where a container is critically under-replicated and we need a good copy of the data as soon as possible. Meanwhile, reconciliation could remain blocked for an unrelated reason. We can provide a way to pause/unpause reconciliation between chunk writes to do reasonably consistent container exports.
+
+#### On block delete
+
+- When SCM sends block delete commands to datanodes, update the merkle tree when the datanode block deleting service runs and processes that block.
+
+    - Merkle tree update should be done before RocksDB update to make sure it is persisted in case of a failure during delete.
+
+    - Merkle tree update should not be done when datanodes first receives the block delete command from SCM, because this command only adds the delete block proto to the container. It does not iterate the blocks to be deleted so we should not add that additional step here.
+
+#### On Container delete for container being reconciled
+
+- Provide a way to interrupt reconciliation and allow container delete to proceed. 
+
+## Backwards Compatibility
+
+Since the scanner can generate the container merkle trees in the background, existing containers created before this feature will still be eligible for reconciliation. These old containers may not have all of their block deletes present in the merkle tree, however, which could cause some false positives about missing blocks on upgrade if one node had already deleted blocks from a container before the upgrade, and another datanode has not yet processed the delete of those blocks.
+
+This can be mitigated by:
+1. Having datanodes delete blocks from their container replica on reconciliation that another replica has marked as deleted.
+2. Having SCM periodically ask datanodes to reconcile otherwise matching containers when there are no other mismatched containers in the cluster.
+
+The sequence of events would look like this. Assume software v1 does not have container reconciliation, but v2 does.
+1. Datanode 1 in v1 deletes block 1 in container 1. This does not leave a tombstone entry because v1 does not have container reconciliation.
+2. Cluster is upgraded to v2.
+3. Reconciliation is triggered between datanodes 1 and 2 for container 1. Datanode 2 has not yet deleted block 1.
+4. Datanode 1 will add block 1 to its container when reconciling with datanode 2.
```

Comment:
If the `delTX` of Datanode 2 is greater than Datanode 1, Datanode 1 can confirm that there is no need to restore Block 1 by deleting the tombstone of Datanode 2.
So maybe you can do this by waiting for all replicas' `delTX` to be greater than a certain value (perhaps the largest `delTX` among all replicas at the beginning of reconciliation)